core/src/main/java/overflowdb/storage/NodesWriter.java

```diff
@@ -0,0 +1,78 @@
+package overflowdb.storage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import overflowdb.Node;
+import overflowdb.NodeDb;
+import overflowdb.NodeRef;
+
+import java.io.IOException;
+import java.util.Spliterator;
+import java.util.stream.StreamSupport;
+
+/**
+ * Persists collections of nodes in bulk to disk. Used either by ReferenceManager (if overflow to disk is enabled),
+ * or alternatively when closing the graph (if storage to disk is enabled).
+ */
+public class NodesWriter {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final boolean isLogInfoEnabled = logger.isInfoEnabled();
+  private static final int LogAfterXSerializedNodes = 100_000;
+  private final NodeSerializer nodeSerializer;
+  private final OdbStorage storage;
+
+  public NodesWriter(NodeSerializer nodeSerializer, OdbStorage storage) {
+    this.nodeSerializer = nodeSerializer;
+    this.storage = storage;
+  }
+
+  /**
+   * writes all references to storage, blocks until complete.
+   */
+  public void writeAndClearBatched(Spliterator<? extends Node> nodes, int estimatedTotalCount) {
+    CountEstimates countEstimates = new CountEstimates();
+
+    StreamSupport.stream(nodes, false).forEach(node -> {
```

Comment:
Can we add a comment that this is sequential, and the reason that it has to be sequential? At least I had a hiccup because I first pattern-matched and assumed parallel stream, searching for races, until I thought to look up the docs for `java.util.stream.StreamSupport.stream`...

Reply:
Actually, I just went through the git history and saw that Denis made the serialization parallel, but made the 'write to storage' sequential. I think he measured the MVStore performance, and given that we're writing to disk, that probably makes sense. Unfortunately he didn't add comments (besides the commit title), but I'll change that. 

Thanks for bringing this up, it triggered me to dig deeper.

Reply:
historical reference: https://github.com/ShiftLeftSecurity/overflowdb/pull/67